### ISSUE TITLE
[Feat] Liveview, captureVideoAsync saves empty file, captureVideo just returns the error

### ISSUE THREAD

**Issue by goooseman** [TypedProject/ts-gphoto2-driver#2]
Dear @Romakita,

Thanks for the great lib. It is a very good replacement of an old node-gphoto2 project. I can help with documentation and maybe some collaboration if I will understand some primary aspects of the library.

Because for now there are some misunderstandings.

What I'm trying to achieve is to view the live preview video file.

What I have:

- Canon 80D
- GPhoto2 2.5.17 installed with brew on Mac
- --capture-movie flag works fine with my camera. Tested with [gphoto2-liveview-example](https://github.com/aqiank/gphoto2-liveview-example).

So first, I'm initialising a camera object. This is easy.

```
import { Camera, closeQuietly } from "@typedproject/gphoto2-driver";
const camera = new Camera();
camera.initialize();
```
Then I'm trying to use captureVideo and getting an error:
```
camera.captureVideo(); // gp_camera_capture returned -6: Unsupported operation
```

Ok, let's try an async method. It works and returns CameraFile object. So let's try to save this file after 5 secs.

```
let video;
try {
  video = await camera.captureVideoAsync();
} catch (e) {
  console.error('Video err', e);
}
console.log('buffer.length', video.buffer.length); // 8
const date = new Date();
const pathToSave = path.join(__dirname, `../.tmp/video-${date.toString()}.mp4`);
setTimeout(async () => {
  console.log('before save buffer.length', video.buffer.length); // 8
  console.log('ref', video.ref()); // undefined
  try {
    const result = await video.saveAsync(pathToSave);
    console.log('saved', result); // 0
  } catch (e) {
    console.error('save err', e);
  }
  video.closeQuietly();
  camera.closeQuietly();
}, 5000);
```
The code works as expected, do not throw any errors at all.
But the result file is 0 bytes.

Plus, what is more interesting: I definitely **do not hear** the sound of mirror shutter, so I believe or it is filming a black screen, or it even do not start to do anything. When I use `gphoto2 --capture-movie --stdout` in the terminal the shutter sound definitely exists.

Looking forward for any help. I can contribute to the library by myself, but for now I can not understand where to dig.

**Comment by Romakita** [comment]
Hi @goooseman 
Your help is welcome :D.

Video and sound aren't supported correctly. I just made a port of the Java version (https://github.com/mvysny/gphoto2-java).

I already tried to implement the LiveView without success.  Example is available here: https://github.com/TypedProject/ts-gphoto2-driver/pull/3
> The problem here, the mirror move as expect but after 1 second, the liveview is interrupted.

For the movie, here is the code used by gphoto2 cli:

```c
int action_camera_capture_movie(GPParams *p, const char *arg) {
    CameraFile *file;
    int r;
    int fd;
    time_t st;
    enum moviemode mm;
    int frames, captured_frames = 0;
    char *xname;
    struct timeval starttime;

    if (p->flags & FLAGS_STDOUT) {
        fd = dup(fileno(stdout));
        xname = "stdout";
    } else {
        fd = open("movie.mjpg", O_WRONLY | O_CREAT, 0660);
        if (fd == -1) {
            cli_error_print(_("Could not open 'movie.mjpg'."));
            return GP_ERROR;
        }
        xname = "movie.mjpg";
    }
    if (!arg) {
        mm = MOVIE_ENDLESS;
        fprintf(stderr, _("Capturing preview frames as movie to '%s'. Press Ctrl-C to abort.\n"), xname);
    } else {
        if (strchr(arg, 's')) {
            sscanf(arg, "%ds", &frames);
            fprintf(stderr, _("Capturing preview frames as movie to '%s' for %d seconds.\n"), xname, frames);
            mm = MOVIE_SECONDS;
            time(&st);
        } else {
            sscanf(arg, "%d", &frames);
            fprintf(stderr, _("Capturing %d preview frames as movie to '%s'.\n"), frames, xname);
            mm = MOVIE_FRAMES;
        }
    }
    CR (gp_file_new_from_fd(&file, fd));
    gettimeofday(&starttime, NULL);
    while (1) {
        const char *mime;
        r = gp_camera_capture_preview(p->camera, file, p->context);
        if (r < 0) {
            cli_error_print(_("Movie capture error... Exiting."));
            break;
        }
        gp_file_get_mime_type(file, &mime);
        if (strcmp(mime, GP_MIME_JPEG)) {
            cli_error_print(_("Movie capture error... Unhandled MIME type '%s'."), mime);
            break;
        }

        captured_frames++;

        if (glob_cancel) {
            fprintf(stderr, _("Ctrl-C pressed ... Exiting.\n"));
            break;
        }
        if (mm == MOVIE_FRAMES) {
            if (!frames--)
                break;
        }
        if (mm == MOVIE_SECONDS) {
            if ((-timediff_now(&starttime)) >= frames * 1000)
                break;
        }
    }
    gp_file_unref(file);

    fprintf(stderr, _("Movie capture finished (%d frames)\n"), captured_frames);
    return GP_OK;
}
```

One interesting thing is the code loop to maintain the process... Maybe we can do the thing ?

I added you to this repo ;) 
See you,
Romain

**Comment by goooseman** [comment]
Hi @Romakita,

Thanks for the info, did not see this PR. I will clone this repo and continue to work in this branch tomorrow. For now my camera's battery is dead and I left the charger at home.

By the way, I've just wrote a small PoC, that it can work using spawn from NodeJS `child_process` inside electron app.

```
const Dicer = require('dicer');
import { spawn } from 'child_process';

const boundary = '--videoboundary';
const dice = new Dicer({boundary: boundary});
dice.on('part', function(part) {
  let frameEncoded = '';
  part.setEncoding('base64');
  part.on('header', console.log);
  part.on('data', (data) => { frameEncoded += data; });
  part.on('end', () => {
    console.log('end', frameEncoded)
    // frameEncoded should be base64 of the image now
    const base64 = "data:image/jpeg;charset=utf-8;base64," + frameEncoded;
    const previewImage = document.getElementById("preview-image") as HTMLImageElement;
    if (previewImage) {
      previewImage.src = base64;
    } else {
      document.getElementById("preview").innerHTML = `<img id="preview-image" src="${base64}" />`;
    }
  });
});

const child = spawn('gphoto2', ['--capture-movie', '--stdout']); // Get the video stream from gphoto2
child.stdout.pipe(dice);
```

**Comment by Romakita** [comment]
Hoo cool :D
I'd just created the PR this morning to share the code with you ;)

**Comment by goooseman** [comment]
(Couldn't wait to test your code, so got back to home to take camera charger).

@Romakita You know, but you code is actually working exactly like gphoto2 cli.

Gphoto2 cli outputs a video file in a strange format (not playable by players I know), so there is a project, which was created to view these video files called [gphoto2-liveview-example](https://github.com/aqiank/gphoto2-liveview-example). (On Mac you need to install sdl2_image and sdl2 using brew to build it)

There are two ways to use it:

1. Pipe it: `gphoto2 --capture-movie --stdout | ./gphoto2-liveview-example`
2. Give it a path to file, which was recorded by gphoto2: `./gphoto2-liveview-example movie.mjpg`

So I took your example (`examples/camera-liveview.js`), changed timeout to 10 secs and make it to output to stdout inside:
``` js
  liveview.on('data', (chunk) => {
    process.stdout.write(chunk);
  });
```

Then I launch it to write stdout to file `node examples/camera-liveview.js > movie.mjpg`.
And then `../gphoto2-liveview-example/gphoto2-liveview-example movie.mjpg` plays the video. The video was a little bit faster, but it works the same with original gphoto2 cli.

So OK, something works.

The strange thing, is that piping to gphoto2-liveview-example does not work.

If I do:
`node examples/camera-liveview.js | ../gphoto2-liveview-example/gphoto2-liveview-example`
It just launches a window with a black screen. 

If I output stdout to console `node examples/camera-liveview.js | cat`, the start of the output seems completely the same with original gphoto2 cli.

I tried to rewrite your code with a while loop - no change. Output to file works, direct piping - not.
I tried to change the number of milliseconds in the interval - no change.


---
By the way, my example from the previous post does not work too. (I didn't test it because of bad battery too).

So there are two things in my plans:

1. Understand why direct piping does not work between our example and gphoto2-liveview-example.
2. Undestand how to parse this binary stream, divide it to frames and output as base64 images.

**Comment by goooseman** [comment]
Ok, now I see the problem.
I read about JPG format [here](https://www.media.mit.edu/pia/Research/deepview/exif.html):
Every JPEG file starts from binary value '0xFFD8', ends by binary value '0xFFD9'

When we are doing `node examples/camera-liveview.js | ../gphoto2-liveview-example/gphoto2-liveview-example`, gphoto2-liveview waits for the start of JPG file (`0xffd8`) (main.c:85)
And this happens OK.

Then it is waiting for the end of the JPG file (`0xffd9`), but this never happens, so it is stuck in the infinite loop.

Why there is no end of the JPG file? Trying to investigate further.

**Comment by goooseman** [comment]
I use `| xxd | grep ffd9` to check for the end of image bytes.
From gphoto2 cli we see a lot of them:
`gphoto2 --capture-movie --stdout | xxd | grep ffd9`
<img width="811" alt="screen shot 2018-08-16 at 18 38 20" src="https://user-images.githubusercontent.com/3503252/44219033-ac66b500-a183-11e8-95a2-118b96a84a2f.png">

With our example there are no:
` node examples/camera-liveview.js | xxd | grep ffd9`

<img width="872" alt="screen shot 2018-08-16 at 18 38 30" src="https://user-images.githubusercontent.com/3503252/44219088-d61fdc00-a183-11e8-9941-4ce6f23b5154.png">

But there are "start of image" bytes:
`node examples/camera-liveview.js | xxd | grep ffd8`

<img width="856" alt="screen shot 2018-08-16 at 18 41 13" src="https://user-images.githubusercontent.com/3503252/44219145-fd76a900-a183-11e8-9a1c-c1941ef09a08.png">


So the only issue left is the "end of image" byte missing.

**Comment by goooseman** [comment]
I was sure that the issue is with this line: `GPhoto2Driver.gp_file_get_mime_type(this.buffer.deref(), mime);`. But when I delete it, rebuild the library, and run the example I still see this image/jpeg as a string in the output. 

So somehow libgphoto2 is returning an `image/jpeg` as a string instead of `0xffd9` at the end of each image when we fire `gp_camera_capture_preview` command.

This is really weird, what can I say.

I have `libgphoto2 2.5.19` installed using homebrew.

I've updated the branch with my `camera-liveview.js` example file.

Can you please confirm, that when you do `node examples/camera-liveview.js | xxd | grep ffd8` you see some strange `image/jpeg` string in the binary of the file?

**Comment by Romakita** [comment]
@goooseman Awesome :) I'll never find this alone.
Ok I'll try ASAP your code.

Thanks for you help :D

**Comment by goooseman** [comment]
One more important thing that I found right now is that `GPhoto2Driver.gp_file_new_from_fd(this.buffer, (process.stdout as any).fd), "gp_file_new_from_fd"` already writes output to stdout.
So no need to write to it in `examples/liveview.js`

**Comment by Romakita** [comment]
You've right. But it's just for test. For the final version, I want use a pipe/buffer/file instead of the process.stdout.fd. I didn't find a way to use anything other than this stdout.

**Comment by Romakita** [comment]
Ok I removed the console.log with the mime in the onTick method. I think, this is the problem (my bad).
LivePreview output correctly ffd8 and ffd9 :)
The branch is up to date now.

I saw this lib https://www.npmjs.com/package/node-sdl2 to implement sdl2 directly in node.js. It could a good idea to use it. no ?

See you

**Comment by goooseman** [comment]
Thank for the help.

Strange, but I already tried to remove console.log and even with your update I do not see ffd9, still see 'image/jpg'. Maybe I have an illegal version of libgphoto2.

I have few questions:

1. What version of libgphoto2 do you have and how did you install it?
2. I see that you reference ../src (not ../lib) folder directly in the example now. How do you launch this example? Typescript does not want to launch .js file
3. I believe we do not need SDL here, because we can show the image in different ways in a web app. What are our plans for an output of this live view? What I was thinking about, we can make it a little bit more high level with some options. We can make an option in a constructor smth like outputType which accepts one of the following: `binary` (returns in data event just the binary), `base64` - returns in data event base64 of each frame. 
4. What do you think about doing an `FPS` option in the constructor? This is how many frames per seconds we wants. We will pass this number to the setInterval (1000 / FPS).

Looking forward for your reply and have a great weekends :)

**Comment by Romakita** [comment]
1. v2.5.18 - If your gphoto2 works fine, normally, the libphoto2 with node.js should work as expected, no ?
2. My bad, I'ts because I use webstorm and it compile automatically the typescript. If you want run example from src, you can run `tsc -w`;)
3. OK ^^ good idea.
4. Good idea too. 

I've updated the branch. Now liveview method accept an options to configuration Liveview class ;)

Have a great  weekends too ;)

**Comment by goooseman** [comment]
@Romakita it is super strange, but I still do not have ffd9, but see image/jpg

Do you run example with 
`
  liveview.on('data', (chunk) => {
    process.stdout.write(chunk);
  });
`
?

What else can be different from your machine (set up) to mine?

**Comment by goooseman** [comment]
So what I've tried to do:

1. Uninstall libghoto2 (`brew uninstall gphoto2 libgphoto2`)
2. Download and compile libgphoto2 from [sources](https://github.com/gphoto/libgphoto2) 
3. Installed mode 8.9.3 using nvm (as in your .nvmrc)
4. Deleted ts-gphoto2-driver folder
5. git clone ts-gphoto2-driver project one more time and cd to it
6. git checkout feat_live_view 
7. npx tsc
8. node examples/camera-liveview.js | xxd | grep ffd8

And still the same: I do not ffd9 after ffd8, but see these image/jpeg:

<img width="1435" alt="screen shot 2018-08-18 at 9 51 23" src="https://user-images.githubusercontent.com/3503252/44296736-cbbd2980-a2cc-11e8-89a1-d01049425e78.png">

**Comment by Romakita** [comment]
Hi @goooseman 

When I run node `examples/camera-liveview.js | xxd | grep ffd9` I have the ffd9 ;)

I added a npm command to facilitate your work, now you able to run `npm run develop`. All examples are linked to `./src`. I hope this command will help you ;).

The branch is up to date :D 

See you,
Romain

**Comment by goooseman** [comment]
@Romakita can you send me  a screenshot of your terminal after `examples/camera-liveview.js | xxd | grep ffd9`
Want to check smth.

**Comment by Romakita** [comment]
Here the screenshot :
<img width="507" alt="image" src="https://user-images.githubusercontent.com/1763311/44306858-929dbb80-a397-11e8-9b79-d09ff468a9b3.png">

**Comment by goooseman** [comment]
@Romakita do you know, how can we stop gphoto2 driver to write to stdout (`GPhoto2Driver.gp_file_new_from_fd`) automatically?

Anyway in a real example, we should return binary or base64 to a stream (`.on('data', callback)`), we should not write to stdout automatically.

But for now I do not understand what I can pass as a second parameter to `GPhoto2Driver.gp_file_new_from_fd` function to stop it to write to stdout.

**Comment by Romakita** [comment]
We have two choices:
1) Give another fd resource to  gp_file_new_from_fd. A kind of buffer for example.
2) Find a way to use gp_file_new_from_handler. But we have to create the CameraFileHandler struct before.

I think create a tmp file isn't a good way for performance reason. But it's relatively easy to implement that:

```typescript
import * as fs from "fs";
import * as path from "path";

const TMP_PATH = path.join(__dirname, "../../.tmp/liveview");

export class Liveview extends EventEmitter {
  private timer: any;
  private buffer: PointerOf<PointerCameraFile>;

  constructor(private camera: PointerCamera, private options: Partial<ILiveviewOptions> = {fps: 100}) {
    super();

    this.buffer = alloc(RefCameraFile) as any;

    this.fd = options.stdout ? (process.stdout as any).fd : fs.openSync(TMP_PATH, "w", 0o777);
    checkCode(GPhoto2Driver.gp_file_new_from_fd(this.buffer, (process.stdout as any).fd), "gp_file_new_from_fd");
  }
}
```

I tried this solution, but not sure if it's work fine.

**Comment by Romakita** [comment]
@Nic2012 I've updated the branch. I removed the stdout. Now the output seems to be good :).

**Comment by goooseman** [comment]
@Romakita thanks for the update.

So now the output inside .tmp/preview.jpg is a valid stream with ffd8 and ffd9 bytes inside. But we have the following issues:

1. camera.initialize(); now fails silently. Before the code change, it was showing an error in console, if the camera is not connected. Now it just finishes the process without any errors.
2. `fs.openSync` fails quietly too, if there are some problems (for example folder does not exists). We should show an error there.
3. It is more common to use `/tmp` folder in the root of the filesystem for the tmp files. Maybe we will use it as default and make options for the user to pass another one?
4. fps param should be the number of shots in a second. So the interval should be (1000 / this.options.fps).


These ones I can fix by myself, if you agree with me.

But there is the biggest issue: libgphoto2 do not clean the file on each "capture_preview". So the file is a buffer of the last frame + all the previous frames. 

And on each tick we have the following code:
```
    fs.readFile(TMP_PATH, {}, (err, data) => {
      this.emit("data", data);
    });
```

Which sends all the contents of the file on each tick. But we need to send only the last frame, not all the frames.

I'll try to think about this problem.

**Comment by goooseman** [comment]
@Romakita we can use smth like `memfs` for this (https://www.npmjs.com/package/memfs)
What do you think?

**Comment by Romakita** [comment]
1) camera.initialize(); now fails silently. Before the code change, it was showing an error in console, if the camera is not connected. Now it just finishes the process without any errors.

Yes my bad. In Garbage.ts the last line should be removed. It catch all errors... 
https://github.com/TypedProject/ts-gphoto2-driver/blob/master/src/components/Garbarge.ts#L37
I'll remove that on another branch and create a new release.

2) fs.openSync fails quietly too, if there are some problems (for example folder does not exists). We should show an error there.

3)  It is more common to use /tmp folder in the root of the filesystem for the tmp files. Maybe we will use it as default and make options for the user to pass another one?

You've right... I forgot this point. You can change that ;). Same for memfs :) 

4) fps param should be the number of shots in a second. So the interval should be (1000 / this.options.fps).

Yep. You've right :)
The problem for the come from the used flag 'w+', we can change that by 'w'. All frame will be removed for each writing process.  https://github.com/TypedProject/ts-gphoto2-driver/blob/feat_live_view/src/components/Liveview.ts#L39

You can change everything if it's works fine ;)

**Comment by goooseman** [comment]
@Romakita ok, I'll work on it

**Comment by goooseman** [comment]
@Romakita If we are using `CameraFile` object, can we some how get access to the binary data of the file without writing it to the disk?

I've tried to use `gp_file_get_data_and_size` (as I saw you have added this function to CameraFile recently), but I get `gp_file_get_data_and_size on GPhoto2Driver doesn't exists` error from the driver.

**Comment by goooseman** [comment]
So what I am thinking about.
Anyway (even without live view) it is a good idea to have an easy way to access binary data inside CameraFile object.

For example, if I make a photo inside my electron application, I would like to get binary data of the photo to render it faster without saving it to fs/reading from fs.

I was looking at `python-gphoto2` project. They have some high-level abstraction `CameraFile.get_data_and_size` which just returns python FileData object. ( [info](https://github.com/jim-easterbrook/python-gphoto2#gp_file_get_data_and_size--camerafileget_data_and_size) / [source](https://github.com/jim-easterbrook/python-gphoto2/blob/424ba5ccca39f5db87f9809c72e0b130fe008c0d/src/gphoto2/file.i))

I want to make the same high-level abstraction in our CameraFile class: `CameraFile.getDataAndSize`. So it will help us to make a liveview without saving file and will be handy in many other use cases.

I see that you've added already this method. Does it work for you?

I'm trying to use it like this:

```
  private async onTick() {
    const data = GPPointerString();
    const size : PointerOf<number> = GPPointer('int');
    await this.camera.callAsync("capture_preview", this.file.pointer, Context.get().pointer);
    this.file.getDataAndSize(data, size);
  }
```

But I get `gp_file_get_data_and_size on GPhoto2Driver doesn't exists` error in the console.

**Comment by Romakita** [comment]
Hum maybe I missed to add the method in driver module.

You have two major directories in this project:
- `components` is dedicated on the features, and hide the complexity of the driver (wrapper of libphoto2 written in C)
- `driver` is dedicated of the driver implementation with ffi and ref library.

When a method is missing like gp_file_get_data_and_size, that means his description is missing in the driver.

So to fix that, you can edit this file https://github.com/TypedProject/ts-gphoto2-driver/blob/master/src/driver/modules/GPFileModule.ts#L12

According to the gphoto2 documentation (http://www.gphoto.org/doc/api/gphoto2-file_8c.html#afe4a2c1685df174f61f20b2986edd93b)

gp_file_get_data_and_size has three parameters:
- CameraFile * file, (JS => a PointerOf CameraFile)
- const char **,  (JS => a string)
- unsigned long int * size, (PointerOf number)

So the description for ffi module should be something like that:

```typescript
export const GPFileModuleDescription = {
  gp_file_new: ["int", [refType(RefCameraFile)]],
  gp_file_new_from_fd: ["int", [refType(RefCameraFile), "int"]],
  gp_file_get_mime_type: ["int", [RefCameraFile, refType(types.CString)]],
  gp_file_ref: ["int", [RefCameraFile]],
  gp_file_clean: ["int", [RefCameraFile]],
  gp_file_save: ["int", [RefCameraFile, types.CString]],
  gp_file_unref: ["int", [RefCameraFile]],
  gp_file_free: ["int", [RefCameraFile]],

  // new method
  gp_file_get_data_and_size["int", [RefCameraFile, refType(types.CString), refType(types.int)]
};
```

I propose you to change the getDataAndSize implementation also. It should probably better the method return directly an Object with data and size. 

```typescript
  public getDataAndSize(): {data: string, size: number} {
       const data = GPPointerString();
       const size : PointerOf<number> = GPPointer('int');

      this.call("get_data_and_size", data, size);
      
      return {data, size};
  }
```
> getMimeType should also follow the same implementation. 

What do you think ? 

See you
Romain

**Comment by goooseman** [comment]
@Romakita thx for the help, will implement this tomorrow and get back to you

**Comment by goooseman** [comment]
@Romakita today I was working on getting binary from CameraFile and on live view functionality.

For a liveview initialiser I have made an encoding param, which can be "binary" | "base64" | "file".

For "binary" we will get binary data of every frame in `.on('data', (data, size))` event.
For "base64" we will get base64 data of every frame in `.on('data', (data, size))` event.
For "file" option we will save the live preview in the file, which will be saved in the path provided by `filePath` param.

But, I have some problems.

1. The most major one: in CameraFile there is a getDataAndSize method. It is a high-level API to get binary of the file.
There is a following code inside:
```ts
    const dataPointer = GPPointerString();
    const sizePointer: PointerOf<number> = GPPointer("int");
    const code = await this.callAsync("get_data_and_size", dataPointer, sizePointer);
```
And if I do `console.log(dataPointer.deref())` after this code I just see `��` in the console. This is the output of live view:

<img width="1018" alt="screen shot 2018-08-21 at 18 58 31" src="https://user-images.githubusercontent.com/3503252/44413518-3e036780-a574-11e8-818d-ff76a29dad3e.png">

So there is no binary at all. 

I did not investigate this issue yet, I will try to do smth tomorrow, but for now I do not understand where the problem can be.

2. After liveview.stop() method and exiting the script I see `malloc: *** error for object 0x102609d00: pointer being freed was not allocated *** set a breakpoint in malloc_error_break to debug` error in the console. Did I forget to close smth?

3. Currently, with the encoding file we get only the last frame, because `capture_preview` method overwrites the contents of the CameraFile. I can rewrite here the code back to `CameraFileFromFd` method, so library will write directly to the file.

I wrote TODO comments in the code to mark places, where I will continue to work.

If you have any thoughts or ideas (especially about the first and second problems or about the code in common), I would be really happy to hear from you 😄

**Comment by Romakita** [comment]
@goooseman Awesome works ;)

To answer you, the gp_file_get_data_and_size is here to implement drivers for specific model. So We can't use this method because it means we need to reimplement gp_camera_capture_preview for each model... gp_camera_capture_preview for CANON is very different of KONICA.

We haven't choice to use gp_camera_capture_preview, write to a file and read the file. 

I updated and cleaned the branch ;). In future if you want update your branch before push, you can use gflow push or gflow rebase.

See you
Romain

**Comment by goooseman** [comment]
1. The changes does not help, I still get � signs instead of binary data for the string after `get_data_and_size` method.

<img width="1025" alt="screen shot 2018-08-22 at 17 48 08" src="https://user-images.githubusercontent.com/3503252/44471051-c4cb4980-a633-11e8-9a74-a05a03c737fc.png">

2. I did not find any information, that gp_file_get_data_and_size and gp_camera_capture_preview methods can be different for different models. For some models, they just do not support capture_preview, but we can do nothing here. I did not get what is our plan? How it is better to implement liveview?

Looking forward for your reply,
Alex.

**Comment by goooseman** [comment]
What I'm more interesting in: why can gp_file_get_data_and_size not be working? It is my first experience with working with C libraries from nodejs, maybe I've implemented something work?

Can you confirm, that you see the same output form this method?

**Comment by goooseman** [comment]
I've read documentation one more time:
http://www.gphoto.org/doc/api/gphoto2-file_8c.html#afe4a2c1685df174f61f20b2986edd93b

And I see that param is `const char ** data`. So it is not a pointer to string. It is a pointer to pointer to string.

I rewrote the code, so it should work now.
Unfortunately, my camera is at home, so I will test this code only in the evening.

**Comment by Romakita** [comment]
Coool :)

**Comment by Romakita** [comment]
Tell when you ready to review the PR ^^. Great job ;)

**Comment by goooseman** [comment]
Hooray 🎉

Finally, I make it to work.

1) about `get_data_and_size`
To get the binary data from it, we need to pass a pointer of `char *` to `get_data_and_size`. Then we need to call `const binary = reinterpret(dataPointer.deref(), sizePointer.deref());` to get binary, where `reinterpret` is `import {reinterpret} from "ref";`
So now we can get binary from any `CameraFile` instance, which can be handy not only for preview.

2) About `malloc: *** error for object 0x102609d00: pointer being freed was not allocated *** set a breakpoint in malloc_error_break to debug` 
LiveView was firing `this.file.closeQuietly()` twice, when the node process was exiting. So I made a check, that this.file exists and make it undefined after first closing.


So for now `node examples/camera-liveview-binary.js | ../gphoto2-liveview-example/gphoto2-liveview-example` is working perfectly, and I can see I live preview from the camera with a good FPS and nearly without any lag.

`node examples/camera-liveview-base64.js` works great too, so we can get base64 of any frame in live mode to render it as an image.

The only thing that is missing now, is that `examples/camera-liveview-file.js` is still writing only the last image in the preview. 
I will rewrite `LiveView.ts` tomorrow with `CameraFileFromFd`, so libgphoto2 will write to file directly while getting the live image.

By the way, please check the current code, if it is OK for you.

**Comment by goooseman** [comment]
Oook, what was done in the morning:

1) For `options.output === 'file'` I reverted back the `CameraFileFromFd` way to write to file, so  libgphoto2 should write directly to the file.
The code now throws an error, if the folder does not exists (previously it was falling silently).
2) Changes in examples: put base64 and binary examples in one file (they are nearly the same) + made an example to test if stopping and then starting back the liveview works.
3) Read the PR, refactored a little bit, fixed some code style errors.

The only problem, that I did not test the morning changes. From the code perspective, it is ready for PR. 
About functionality, I will test writing to file and multiple starts of liveview from the same liveview object in the evening.
